ibis/pyspark/compiler.py

```diff
@@ -1011,3 +1058,296 @@ def compile_cumulative_max(t, expr, scope, *, window, **kwargs):
     return _handle_cumulative_operation(
         t, expr, scope, F.max, window=window, **kwargs
     )
+
+
+# -------------------------- Temporal Operations ----------------------------
+
+# Ibis value to PySpark value
+_time_unit_mapping = {
+    'Y': 'year',
+    'Q': 'quarter',
+    'M': 'month',
+    'W': 'week',
+    'D': 'day',
+    'h': 'hour',
+    'm': 'minute',
+    's': 'second'
+}
+
+
+@compiles(ops.Date)
+def compile_date(t, expr, scope, **kwargs):
+    op = expr.op()
+    src_column = t.translate(op.arg, scope)
+    return F.to_date(src_column).cast('timestamp')
+
+
+def _extract_component_from_datetime(t, expr, scope, extract_fn, **kwargs):
+    op = expr.op()
+    date_col = t.translate(op.arg, scope)
+    return extract_fn(date_col)
+
+
+@compiles(ops.ExtractYear)
+def compile_extract_year(t, expr, scope, **kwargs):
+    return _extract_component_from_datetime(t, expr, scope, F.year, **kwargs)
+
+
+@compiles(ops.ExtractMonth)
+def compile_extract_month(t, expr, scope, **kwargs):
+    return _extract_component_from_datetime(t, expr, scope, F.month, **kwargs)
+
+
+@compiles(ops.ExtractDay)
+def compile_extract_day(t, expr, scope, **kwargs):
+    return _extract_component_from_datetime(
+        t, expr, scope, F.dayofmonth, **kwargs
+    )
+
+
+@compiles(ops.ExtractHour)
+def compile_extract_hour(t, expr, scope, **kwargs):
+    return _extract_component_from_datetime(t, expr, scope, F.hour, **kwargs)
+
+
+@compiles(ops.ExtractMinute)
+def compile_extract_minute(t, expr, scope, **kwargs):
+    return _extract_component_from_datetime(t, expr, scope, F.minute, **kwargs)
+
+
+@compiles(ops.ExtractSecond)
+def compile_extract_second(t, expr, scope, **kwargs):
+    return _extract_component_from_datetime(t, expr, scope, F.second, **kwargs)
+
+
+@compiles(ops.ExtractMillisecond)
+def compile_extract_millisecond(t, expr, scope, **kwargs):
+    raise com.UnsupportedOperationError(
+        'PySpark backend does not support extracting milliseconds.'
+    )
+
+
+@compiles(ops.DateTruncate)
+def compile_date_truncate(t, expr, scope, **kwargs):
+    op = expr.op()
+
+    try:
+        unit = _time_unit_mapping[op.unit]
+    except KeyError:
+        raise com.UnsupportedOperationError(
+            '{!r} unit is not supported in timestamp truncate'.format(op.unit)
+        )
+
+    src_column = t.translate(op.arg, scope)
+    return F.date_trunc(unit, src_column)
+
+
+@compiles(ops.TimestampTruncate)
```

Comment:
you can just make the above dispatcher match on either type. so `@compiles((ops.DateTruncate, ops.TimestampTruncate))`

Reply:
Hmm when I try this, I get `XFAIL ibis/tests/all/test_temporal.py::test_timestamp_truncate[PySpark-Y]
  reason: PySpark: No translation rule for <class 'ibis.expr.operations.TimestampTruncate'>`.

Reply:
odd. okay, fine to leave as is